tracer/test/Datadog.Trace.Tests/Telemetry/TelemetryCircuitBreakerTests.cs

```diff
@@ -0,0 +1,120 @@
+        public void OnInitialError_TreatsAsTransientAndSavesDataForLater(int errorType)
+        {
+            var config = new ConfigTelemetryData();
+            var deps = Array.Empty<DependencyTelemetryData>();
+            var integrations = Array.Empty<IntegrationTelemetryData>();
+
+            var telemetryPushResult = _circuitBreaker.Evaluate((TelemetryPushResult)errorType, config, deps, integrations);
+
+            telemetryPushResult.Should().Be(TelemetryPushResult.TransientFailure);
+            _circuitBreaker.PreviousConfiguration.Should().BeSameAs(config);
+            _circuitBreaker.PreviousDependencies.Should().BeSameAs(deps);
+            _circuitBreaker.PreviousIntegrations.Should().BeSameAs(integrations);
+        }
+
+        [Fact]
+        public void OnMultipleInitialFatalError_ReturnsFatal()
+        {
+            var config = new ConfigTelemetryData();
+            var deps = Array.Empty<DependencyTelemetryData>();
+            var integrations = Array.Empty<IntegrationTelemetryData>();
+
+            _circuitBreaker.Evaluate(TelemetryPushResult.FatalError, config, deps, integrations);
+            var telemetryPushResult = _circuitBreaker.Evaluate(TelemetryPushResult.FatalError, config, deps, integrations);
+
+            telemetryPushResult.Should().Be(TelemetryPushResult.FatalError);
+            _circuitBreaker.PreviousConfiguration.Should().BeNull();
+            _circuitBreaker.PreviousDependencies.Should().BeNull();
+            _circuitBreaker.PreviousIntegrations.Should().BeNull();
+        }
+
+        [Fact]
+        public void OnMultipleFatalErrorAfterSuccess_ReturnsTransient()
+        {
+            var config = new ConfigTelemetryData();
+            var deps = Array.Empty<DependencyTelemetryData>();
+            var integrations = Array.Empty<IntegrationTelemetryData>();
+
+            _circuitBreaker.Evaluate(TelemetryPushResult.Success, config, deps, integrations);
+            var firstResult = _circuitBreaker.Evaluate(TelemetryPushResult.FatalError, config, deps, integrations);
+            var secondResult = _circuitBreaker.Evaluate(TelemetryPushResult.FatalError, config, deps, integrations);
+
+            firstResult.Should().Be(TelemetryPushResult.TransientFailure);
+            secondResult.Should().Be(TelemetryPushResult.TransientFailure);
+        }
+
+        [Fact]
+        public void OnMultipleTransientErrorsAfterSuccess_ReturnsFatal()
+        {
+            var config = new ConfigTelemetryData();
+            var deps = Array.Empty<DependencyTelemetryData>();
+            var integrations = Array.Empty<IntegrationTelemetryData>();
+
+            _circuitBreaker.Evaluate(TelemetryPushResult.Success, config, deps, integrations);
+            for (var i = 0; i < TelemetryCircuitBreaker.MaxTransientErrors - 1; i++)
+            {
+                var result = _circuitBreaker.Evaluate(TelemetryPushResult.FatalError, config, deps, integrations);
```

Comment:
Should the results passed to `_circuitBreaker.Evaluate` in this test be `TelemetryPushResult.TransientFailure`?

Reply:
We should probably test both. Effectively, the `TelemetryCircuitBreaker` treats `FatalError` and `TransientError` the _same_ as long as there's been previously bee a successful push result. 

I've updated this to be a theory that tests with both types of error in https://github.com/DataDog/dd-trace-dotnet/pull/2241/commits/0c155b8f7cc7b18d61a2fed492afb8100481e6fb, but the logic around this is also up for discussion of course 🙂

Reply:
The logic looked fine to me, I just didn't know which TelemetryPushResult should be tested, so the change you made looks good 👍🏼